### PR TITLE
Fix rocAL Tensor build issues - Metadata and parameter

### PR DESCRIPTION
**Pull request by fiona-gladwin**
- Fix warnings in meta_data.h file
- Fix build issues in meta nodes.
- Change parameter crop to use ROI instead of width and height vectors.

### CODE REVIEW

File: rocAL/rocAL/include/parameters/parameter_crop.h [hunk 1]
```diff
@@ -50,18 +50,15 @@ class CropParam
...
-        in_width  = in_width_;
-        in_height = in_height_;
+        in_roi = roi;
     }
     void set_random() {_random = true;}
     void set_fixed_crop(float anchor_x, float anchor_y) { _is_fixed_crop = true; _random = false; _crop_anchor[0] = anchor_x; _crop_anchor[1] = anchor_y;}
     void set_x_drift_factor(Parameter<float>* x_drift);
     void set_y_drift_factor(Parameter<float>* y_drift);
-    std::vector<uint32_t> in_width, in_height;
+    const RocalROI * in_roi;
```
Comment (sampath1117) [inline rocAL/rocAL/include/parameters/parameter_crop.h:61]: Please remove the space after * in L61

File: rocAL/rocAL/source/meta_data/meta_node_crop.cpp [hunk 2]
```diff
@@ -40,8 +40,8 @@ void CropMetaNode::update_parameters(pMetaDataBatch input_meta_data, pMetaDataBa
     _crop_height = _meta_crop_param->croph_arr;
     _x1 = _meta_crop_param->x1_arr;
     _y1 = _meta_crop_param->y1_arr;
-    _input_width_val = _meta_crop_param->in_width;
-    _input_height_val = _meta_crop_param->in_height;
+    // _input_width_val = _meta_crop_param->in_width;   TODO - Commenting for now to be replaced with ROI
```
Comment (sampath1117) [inline rocAL/rocAL/source/meta_data/meta_node_crop.cpp:43]: This change is going to be updated in next PR's?

File: rocAL/rocAL/source/meta_data/meta_node_crop_mirror_normalize.cpp [hunk 3]
```diff
@@ -45,8 +45,8 @@ void CropMirrorNormalizeMetaNode::update_parameters(pMetaDataBatch input_meta_da
     _dstImgHeight = _meta_crop_param->croph_arr;
     _x1 = _meta_crop_param->x1_arr;
     _y1 = _meta_crop_param->y1_arr;
-    _src_width = _node->get_src_width();
-    _src_height = _node->get_src_height();
+    // _src_width = _node->get_src_width(); // TODO - Commenting for now to be replaced with ROI
+    // _src_height = _node->get_src_height();
```
Comment (sampath1117) [inline rocAL/rocAL/source/meta_data/meta_node_crop_mirror_normalize.cpp:49]: Please check if this changed is intended

File: rocAL/rocAL/source/parameters/parameter_crop.cpp [hunk 4]
```diff
@@ -57,8 +57,6 @@ void CropParam::array_init()
     croph_arr_val.resize(batch_size);
     x2_arr_val.resize(batch_size);
     y2_arr_val.resize(batch_size);
-    in_width.resize(batch_size);
-    in_height.resize(batch_size);
```
Comment (sampath1117) [inline rocAL/rocAL/source/parameters/parameter_crop.cpp:61]: Please check if this deletion is intended

File: rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp [hunk 5]
```diff
@@ -50,36 +50,36 @@ void RocalCropParam::fill_crop_dims() {
...
             // Evaluating user given crop
-            cropw_arr_val[img_idx] = (crop_w <= in_width[img_idx] && crop_w > 0) ? crop_w : in_width[img_idx];
-            croph_arr_val[img_idx] = (crop_h <= in_height[img_idx] && crop_h > 0) ? crop_h : in_height[img_idx];
+            cropw_arr_val[img_idx] = (crop_w <= in_roi[img_idx].x2 && crop_w > 0) ? crop_w : in_roi[img_idx].x2;
+            croph_arr_val[img_idx] = (crop_h <= in_roi[img_idx].y2 && crop_h > 0) ? crop_h : in_roi[img_idx].y2;
             if (_is_fixed_crop) {
-                x1_arr_val[img_idx] = static_cast<size_t>(std::nearbyintf(_crop_anchor[0] * (in_width[img_idx] - cropw_arr_val[img_idx])));
-                y1_arr_val[img_idx] = static_cast<size_t>(std::nearbyintf(_crop_anchor[1] * (in_height[img_idx] - croph_arr_val[img_idx])));
+                x1_arr_val[img_idx] = static_cast<size_t>(std::nearbyintf(_crop_anchor[0] * (in_roi[img_idx].x2 - cropw_arr_val[img_idx])));
+                y1_arr_val[img_idx] = static_cast<size_t>(std::nearbyintf(_crop_anchor[1] * (in_roi[img_idx].y2 - croph_arr_val[img_idx])));
```
Comment (sampath1117) [inline rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp:57]: Please check if we need to validate x1_arr_val and y1_arr_val also after this computation to check if those are valid or not

File: rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp [hunk 6]
```diff
@@ -50,36 +50,36 @@ void RocalCropParam::fill_crop_dims() {
...
         } else {
           float crop_h_factor_, crop_w_factor_, x_drift, y_drift;
           crop_height_factor->renew();
           crop_h_factor_ = crop_height_factor->get();
           crop_width_factor->renew();
           crop_w_factor_ = crop_width_factor->get();
-          cropw_arr_val[img_idx] = static_cast<size_t> (crop_w_factor_ * in_width[img_idx]);
-          croph_arr_val[img_idx] = static_cast<size_t> (crop_h_factor_ * in_height[img_idx]);
+          cropw_arr_val[img_idx] = static_cast<size_t> (crop_w_factor_ * in_roi[img_idx].x2);
+          croph_arr_val[img_idx] = static_cast<size_t> (crop_h_factor_ * in_roi[img_idx].y2);
```
Comment (sampath1117) [inline rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp:69]: Please remove the space after static_cast<size_t>

File: rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp [hunk 7]
```diff
@@ -50,36 +50,36 @@ void RocalCropParam::fill_crop_dims() {
...
           x_drift_factor->renew();
           y_drift_factor->renew();
           y_drift_factor->renew();
```
Comment (sampath1117) [inline rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp:72]: Why are we renewing y_drift factor 2 times here?

File: rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp [hunk 8]
```diff
@@ -50,36 +50,36 @@ void RocalCropParam::fill_crop_dims() {
...
-          y1_arr_val[img_idx] = static_cast<size_t>(y_drift * (in_height[img_idx] - croph_arr_val[img_idx]));
+          x1_arr_val[img_idx] = static_cast<size_t>(x_drift * (in_roi[img_idx].x2 - cropw_arr_val[img_idx]));
+          y1_arr_val[img_idx] = static_cast<size_t>(y_drift * (in_roi[img_idx].y2 - croph_arr_val[img_idx]));
         }
         x2_arr_val[img_idx] = x1_arr_val[img_idx] + cropw_arr_val[img_idx];
         y2_arr_val[img_idx] = y1_arr_val[img_idx] + croph_arr_val[img_idx];
         // Evaluating the crop
-        x2_arr_val[img_idx] = std::min(x2_arr_val[img_idx], in_width[img_idx]);
-        y2_arr_val[img_idx] = std::min(y2_arr_val[img_idx], in_height[img_idx]);
+        x2_arr_val[img_idx] = std::min(x2_arr_val[img_idx], in_roi[img_idx].x2);
```
Comment (sampath1117) [inline rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp:81]: Please check if x1 and y1 also needs to be validated here

File: rocAL/rocAL/source/meta_data/coco_meta_data_reader.cpp [hunk 9]
```diff
@@ -334,7 +334,7 @@ void COCOMetaDataReader::read_all(const std::string &path)
                     bb_labels.push_back(label);
                     polygon_count.push_back(polygon_size);
                     vertices_count.push_back(vertices_array);
-                    add(itr->second, bb_coords, bb_labels, image_size, mask, polygon_count, vertices_count);
+                    add(file_name, bb_coords, bb_labels, image_size, mask, polygon_count, vertices_count);
```
Comment (sampath1117) [inline rocAL/rocAL/source/meta_data/coco_meta_data_reader.cpp:337]: Please check if this change is intended

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 10]
```diff
@@ -25,6 +25,7 @@ THE SOFTWARE.
 #include <utility>
 #include <vector>
 #include <memory>
+#include <cstring>
```
Comment (sampath1117) [inline rocAL/rocAL/include/meta_data/meta_data.h:28]: Where is this new include used?

File: rocAL/rocAL/include/parameters/parameter_crop.h [hunk 11]
```diff
@@ -50,18 +50,15 @@ class CropParam
...
         y_drift_factor     = default_y_drift_factor();
     }
-    void set_image_dimensions( const std::vector<uint32_t>& in_width_, const std::vector<uint32_t>& in_height_)
+    void set_image_dimensions(RocalROI *roi)
     {
-        if(in_width_.size() != in_width.size() || in_height.size() != in_height_.size())
-            THROW("wrong input width = "+ TOSTR(in_width.size())+" or height size = "+TOSTR(in_height_.size()))
-        in_width  = in_width_;
-        in_height = in_height_;
+        in_roi = roi;
```
Comment (sampath1117) [inline rocAL/rocAL/include/parameters/parameter_crop.h:55]: Please add a validation check stating if roi is a NULL pointer
we should add a error statement to handle it

**Review by sampath1117 — COMMENTED** [review]
@fiona-gladwin 
Please address review comments

File: rocAL/rocAL/source/meta_data/meta_node_crop.cpp (hunk 2 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/meta_data/meta_node_crop.cpp:43]: Yes

File: rocAL/rocAL/source/meta_data/meta_node_crop_mirror_normalize.cpp (hunk 3 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/meta_data/meta_node_crop_mirror_normalize.cpp:49]: Yes, will be fixed in future PR

File: rocAL/rocAL/source/parameters/parameter_crop.cpp (hunk 4 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/parameters/parameter_crop.cpp:61]: Yes its intended

File: rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp (hunk 6 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp:69]: Done

File: rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp (hunk 7 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp:72]: Done

File: rocAL/rocAL/include/meta_data/meta_data.h (hunk 10 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/include/meta_data/meta_data.h:28]: memcpy is used here hence this is included

File: rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp (hunk 8 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp:81]: It is validated in the code above these lines please check

File: rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp (hunk 5 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/parameters/parameter_rocal_crop.cpp:57]: Done

File: rocAL/rocAL/source/meta_data/coco_meta_data_reader.cpp (hunk 9 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/meta_data/coco_meta_data_reader.cpp:337]: Yes

**Comment by fiona-gladwin** [comment]
@sampath1117 and @shobana-mcw I have resolved all PR comments please check

**Comment by fiona-gladwin** [comment]
Merged

**Pull request closed without merging**